src/feature/hs/hs_cell.c

```diff
@@ -853,17 +853,46 @@ hs_cell_parse_introduce2(hs_cell_introduce2_data_t *data,
 
   /* First bytes of the ENCRYPTED section are the client public key (they are
    * guaranteed to exist because of the length check above). We are gonna use
-   * the client public key to compute the ntor keys and decrypt the payload: */
-  memcpy(&data->client_pk.public_key, encrypted_section, CURVE25519_PUBKEY_LEN);
-
-  /* Get the right INTRODUCE2 ntor keys and verify the cell MAC */
-  intro_keys = get_introduce2_keys_and_verify_mac(data, encrypted_section,
-                                                  encrypted_section_len);
-  if (!intro_keys) {
-    log_info(LD_REND, "Could not get valid INTRO2 keys on circuit %u "
-             "for service %s", TO_CIRCUIT(circ)->n_circ_id,
-             safe_str_client(service->onion_address));
-    goto done;
+   * the client public key to compute the ntor keys and decrypt the payload:
+   */
+  memcpy(&data->client_pk.public_key, encrypted_section,
+         CURVE25519_PUBKEY_LEN);
+
+  /* If we are configured as an Onion Balance instance, we need to try
+   * validation with the configured master public keys given in the config
+   * file. This is because the master identity key and the blinded key is put
+   * in the INTRODUCE2 cell by the client thus it will never validate with
+   * this instance public key. */
+  if (service->config.ob_master_pubkeys) {
+    hs_cell_introduce2_data_t new_data;
+
+    SMARTLIST_FOREACH_BEGIN(service->config.ob_master_pubkeys,
```

Comment:
I think we might need to iterate over the three possible timeperiods (-1, 0, +1) as we discussed in jitsi. Consider this:

Client has the 23:00 consensus and backend service has the 01:00 consensus. This means that Client is in TP #N-1 and Service is in TP #N. Hence when it does `hs_get_time_period_num(0)` here it will not synch with what the client does.

What do you think?